ethcore/src/blockchain/blockchain.rs

```diff
@@ -527,165 +416,225 @@ impl BlockChain {
 		let mut best_block = self.best_block.write().unwrap();
-		if let Some(b) = update.new_best {
-			*best_block = b;
+		match update.info.location {
+			BlockLocation::Branch => (),
+			_ => {
+				*best_block = BestBlock {
+					hash: update.info.hash,
+					number: update.info.number,
+					total_difficulty: update.info.total_difficulty
+				};
+			}
+		}
+
+		let mut write_hashes = self.block_hashes.write().unwrap();
+		for (number, hash) in &update.block_hashes {
+			batch.put_extras(number, hash);
+			write_hashes.remove(number);
 		}
 
-		// update details cache
 		let mut write_details = self.block_details.write().unwrap();
-		write_details.remove(&update.details.parent);
-		write_details.insert(update.hash.clone(), update.details);
-		self.note_used(CacheID::Block(update.hash));
+		for (hash, details) in update.block_details.into_iter() {
+			batch.put_extras(&hash, &details);	
+			write_details.insert(hash, details);
+		}
+
+		let mut write_receipts = self.block_receipts.write().unwrap();
+		for (hash, receipt) in &update.block_receipts {
+			batch.put_extras(hash, receipt);
+			write_receipts.remove(hash);
+		}
+
+		let mut write_txs = self.transaction_addresses.write().unwrap();
+		for (hash, tx_address) in &update.transactions_addresses {
+			batch.put_extras(hash, tx_address);
+			write_txs.remove(hash);
+		}
 
-		// update blocks blooms cache
 		let mut write_blocks_blooms = self.blocks_blooms.write().unwrap();
-		for bloom_hash in &update.bloom_hashes {
+		for (bloom_hash, blocks_bloom) in &update.blocks_blooms {
+			batch.put_extras(bloom_hash, blocks_bloom);
 			write_blocks_blooms.remove(bloom_hash);
 		}
 
 		// update extras database
-		self.extras_db.write(update.batch).unwrap();
+		self.extras_db.write(batch).unwrap();
 	}
 
-	/// Transforms block into WriteBatch that may be written into database
-	/// Additionally, if it's new best block it returns new best block object.
-	fn block_to_extras_update(&self, bytes: &[u8], receipts: Vec<Receipt>) -> ExtrasUpdate {
-		// create views onto rlp
-		let block = BlockView::new(bytes);
+	/// Get inserted block info which is critical to preapre extras updates.
+	fn block_info(&self, block_bytes: &[u8]) -> BlockInfo {
+		let block = BlockView::new(block_bytes);
 		let header = block.header_view();
-
-		// prepare variables
 		let hash = block.sha3();
-		let mut parent_details = self.block_details(&header.parent_hash()).expect(format!("Invalid parent hash: {:?}", header.parent_hash()).as_ref());
+		let number = header.number();
+		let parent_hash = header.parent_hash();
+		let parent_details = self.block_details(&parent_hash).expect(format!("Invalid parent hash: {:?}", parent_hash).as_ref());
 		let total_difficulty = parent_details.total_difficulty + header.difficulty();
 		let is_new_best = total_difficulty > self.best_block_total_difficulty();
+
+		BlockInfo {
+			hash: hash,
+			number: number,
+			total_difficulty: total_difficulty,
+			location: match is_new_best {
```

Comment:
consider rewriting as:
```
if is_new_best {
	// on new best block we need to make sure that all ancestors
	// are moved to "canon chain"
	// find the route between old best block and the new one
	let best_hash = self.best_block_hash();
	let route = self.tree_route(best_hash, parent_hash);

	assert_eq!(number, parent_details.number + 1);

	match route.blocks.len() {
		0 => BlockLocation::CanonChain,
		_ => BlockLocation::BranchBecomingCanonChain {
			ancestor: route.ancestor,
			route: route.blocks.into_iter().skip(route.index).collect()
		}
	}
} else { BlockLocation::Branch }
```
(saves one line and avoids the `true` and `false` symbols).